secure/secure.sh

```diff
@@ -0,0 +1,403 @@
+  chown -R yarn:hadoop $HADOOP_CONF_DIR/ssl
+  chmod 755 $HADOOP_CONF_DIR/ssl
+  chmod 440 $HADOOP_CONF_DIR/ssl/keystore.jks
+  chmod 444 $HADOOP_CONF_DIR/ssl/truststore.jks
+}
+
+function set_property_ssl_xml() {
+  set_property_in_xml $HADOOP_CONF_DIR/ssl-$1.xml $2 $3
+}
+
+function config_ssl_xml() {
+  cp $HADOOP_CONF_DIR/ssl-$1.xml.example $HADOOP_CONF_DIR/ssl-$1.xml
+  set_property_ssl_xml $1 "ssl.$1.truststore.location" "$HADOOP_CONF_DIR/ssl/truststore.jks"
+  set_property_ssl_xml $1 "ssl.$1.truststore.password" "$keystore_password"
+  set_property_ssl_xml $1 "ssl.$1.truststore.type" "jks"
+  set_property_ssl_xml $1 "ssl.$1.keystore.location" "$HADOOP_CONF_DIR/ssl/keystore.jks"
+  set_property_ssl_xml $1 "ssl.$1.keystore.password" "$keystore_password"
+  set_property_ssl_xml $1 "ssl.$1.keystore.keypassword" "$keystore_password"
+  set_property_ssl_xml $1 "ssl.$1.keystore.type" "jks"
+}
+
+function set_property_hive_site() {
+  set_property_in_xml "/etc/hive/conf/hive-site.xml" $1 $2
+}
+
+function secure_hive() {
+  if [[ $ROLE == 'Master' ]]; then
+    kadmin -p root -w $root_principal_password -q "addprinc -randkey hive/$FQDN"
+    kadmin -p root -w $root_principal_password -q "xst -k hive.keytab hive/$FQDN"
+    mv ./hive.keytab /etc/hive/conf/
+    chown hive:hive /etc/hive/conf/hive.keytab
+    chmod 400 /etc/hive/conf/hive.keytab
+    # Add the user 'hive' to group 'hadoop' so it can access the keystore files
+    usermod -a -G hadoop hive
+    # HiveServer2
+    set_property_hive_site 'hive.server2.authentication' 'KERBEROS'
+    set_property_hive_site 'hive.server2.authentication.kerberos.principal' "hive/_HOST@$REALM"
+    set_property_hive_site 'hive.server2.authentication.kerberos.keytab' '/etc/hive/conf/hive.keytab'
+    set_property_hive_site 'hive.server2.thrift.sasl.qop' 'auth-conf'
+    # Hive Metastore
+    set_property_hive_site 'hive.metastore.sasl.enabled' 'true'
+    set_property_hive_site 'hive.metastore.kerberos.principal' "hive/_HOST@$REALM"
+    set_property_hive_site 'hive.metastore.kerberos.keytab.file' '/etc/hive/conf/hive.keytab'
+    # WebUI SSL
+    set_property_hive_site 'hive.server2.webui.use.ssl' 'true'
+    set_property_hive_site 'hive.server2.webui.keystore.path' "$HADOOP_CONF_DIR/ssl/keystore.jks"
+    set_property_hive_site 'hive.server2.webui.keystore.password' $keystore_password
+  fi
+}
+
+function secure_spark() {
+  if [[ $ROLE == 'Master' ]]; then
+    kadmin -p root -w $root_principal_password -q "addprinc -randkey spark/$FQDN"
+    kadmin -p root -w $root_principal_password -q "xst -k spark.keytab spark/$FQDN"
+    mv ./spark.keytab /etc/spark/conf/
+    chown spark:spark /etc/spark/conf/spark.keytab
+    chmod 400 /etc/spark/conf/spark.keytab
+    # Add the user 'spark' to group 'hadoop' so it can access the keystore files
+    usermod -a -G hadoop spark
+    cat << EOF >> /etc/spark/conf/spark-env.sh
+SPARK_HISTORY_OPTS="-Dspark.history.kerberos.enabled=true
```

Comment:
Just in case there are other opts already defined, `SPARK_HISTORY_OPTS="$SPARK_HISTORY_OPTS ..."`

Reply:
Done.